soh/soh/ActorDB.cpp

```diff
@@ -0,0 +1,597 @@
+    newEntry.entry.id = index;
+
+    assert(!newEntry.entry.valid);
+
+    nextFreeId = std::max(nextFreeId, index + 1);
+
+    nameTable[name] = newEntry.entry.id;
+    newEntry.setName(name);
+    newEntry.setDesc(desc);
+
+    newEntry.entry.valid = true;
+
+    return newEntry;
+}
+
+// Adds an actor with the legacy ActorInit struct. The id is provided in it.
+ActorDB::Entry& ActorDB::Add(const std::string& name, const std::string& desc, const ActorInit& init) {
+    Entry& entry = Add(name, desc, init.id);
+
+    entry.entry.category = init.category;
+    entry.entry.flags = init.flags;
+    entry.entry.objectId = init.objectId;
+    entry.entry.instanceSize = init.instanceSize;
+    entry.entry.init = init.init;
+    entry.entry.destroy = init.destroy;
+    entry.entry.update = init.update;
+    entry.entry.draw = init.draw;
+    entry.entry.reset = init.reset;
+
+    return entry;
+}
+
+// Adds an actor with the new ActorDBInit struct. The id assigned to the actor is dynamic. Use the return Entry or RetrieveId to get it.
+ActorDB::Entry& ActorDB::Add(const ActorDBInit& init) {
+    Entry& entry = Add(init.name, init.desc, nextFreeId);
+
+    entry.entry.category = init.category;
+    entry.entry.flags = init.flags;
+    entry.entry.objectId = init.objectId;
+    entry.entry.instanceSize = init.instanceSize;
+    entry.entry.init = init.init;
+    entry.entry.destroy = init.destroy;
+    entry.entry.update = init.update;
+    entry.entry.draw = init.draw;
+    entry.entry.reset = init.reset;
+
+    return entry;
+}
+
+// Get the ActorDB::Enty for the given actor id.
+ActorDB::Entry& ActorDB::Retrieve(const int id) {
+    static Entry invalid;
+    if ((id < 0) || (id >= db.size())) {
+        return invalid;
+    }
+    return db[id];
+}
+
+// Get the id for a given actor by name.
+int ActorDB::RetrieveId(const std::string& name) {
+    auto entry = nameTable.find(name);
+    if (entry == nameTable.end()) {
+        return -1;
+    }
+
+    return entry->second;
+}
+
+ActorDB::Entry::Entry() {
+    entry.name = nullptr;
+    entry.desc = nullptr;
+    entry.valid = false;
+    entry.id = 0;
+    entry.category = 0;
+    entry.flags = 0;
+    entry.objectId = 0;
+    entry.instanceSize = 0;
+    entry.init = nullptr;
+    entry.destroy = nullptr;
+    entry.update = nullptr;
+    entry.draw = nullptr;
+    entry.reset = nullptr;
+    entry.numLoaded = 0;
+}
+
+ActorDB::Entry::Entry(const Entry& other) {
+    entry = other.entry;
+    setName(other.name);
+    setDesc(other.desc);
+}
+
+ActorDB::Entry& ActorDB::Entry::operator=(const Entry& other) {
+    entry = other.entry;
+    setName(other.name);
+    setDesc(other.desc);
+    return *this;
+}
+
+void ActorDB::Entry::setName(const std::string& newName) {
+    name = newName;
+    entry.name = name.c_str();
+}
+
+void ActorDB::Entry::setDesc(const std::string& newDesc) {
+    desc = newDesc;
+    entry.desc = desc.c_str();
+}
+
+extern "C" ActorDBEntry* ActorDB_Retrieve(const int id) {
```

Comment:
Is ActorDB the **only** way now to interact with Actors? If so, I would say let's name it:
```suggestion
extern "C" ActorDBEntry* GetActorEntry(const int id) {
```

if not then I'd say let's go for

```suggestion
extern "C" ActorDBEntry* ActorDB_GetEntry(const int id) {
```

Reply:
It is the only place to interact with certain information about actors, mainly for spawning them. But, I don't think `GetActorEntry` is appropriate. This is `extern "C"`, so it is in the global namespace, and so I would like it to be namescaped as best we can. The Namespace_Name pattern is common, see `OTRGlobals.h` for a whole list of examples.